switch/ogre3d/ogre-switch.patch

```diff
@@ -0,0 +1,690 @@
++
++THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
++IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
++FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
++AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
++LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
++OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
++THE SOFTWARE.
++-----------------------------------------------------------------------------
++*/
++
++#include "OgreRoot.h"
++#include "OgreException.h"
++
++#include "OgreSwitchEGLSupport.h"
++#include "OgreSwitchEGLWindow.h"
++#include "OgreViewport.h"
++
++#include <switch.h>
++
++namespace Ogre {
++    SwitchEGLWindow::SwitchEGLWindow(SwitchEGLSupport *glsupport)
++        : EGLWindow(glsupport)
++    {
++        mClosed = true;
++    }
++
++    void SwitchEGLWindow::getLeftAndTopFromNativeWindow( int & left, int & top, uint width, uint height )
++    {
++        // We are always fixed at (0, 0) for the Nintendo Switch.
++        left = top = 0;
++    }
++
++    void SwitchEGLWindow::initNativeCreatedWindow(const NameValuePairList *miscParams)
++    {
++    }
++
++    void SwitchEGLWindow::createNativeWindow( int &left, int &top, uint &width, uint &height, String &title )
++    {
++    }
++
++    void SwitchEGLWindow::reposition( int left, int top )
++    {
++    }
++
++    void SwitchEGLWindow::resize(uint width, uint height)
++    {
++    }
++
++    void SwitchEGLWindow::windowMovedOrResized()
++    {
++    }
++
++    void SwitchEGLWindow::switchFullScreen(bool fullscreen)
++    {
++
++    }
++
++    void SwitchEGLWindow::create(const String& name, uint width, uint height,
++                               bool fullScreen, const NameValuePairList *miscParams)
++    {
++        // Most of these are fixed for the Nintendo Switch.
++        mName = name;
++        mLeft = 0;
++        mTop = 0;
++        mIsFullScreen = true;
++        mIsExternal = false;
++        mHwGamma = false;
++        mWidth = 1920;
++        mHeight = 1080;
++
++        // Create the window and GL context.
++        _notifySurfaceDestroyed();
++        _notifySurfaceCreated(nwindowGetDefault(), 0);
++    }
++
++    void SwitchEGLWindow::_notifySurfaceDestroyed()
++    {
++        if(mClosed)
++            return;
++
++        mContext->setCurrent();
++        mContext->_destroyInternalResources();
++
++        eglDestroySurface(mEglDisplay, mEglSurface);
++        EGL_CHECK_ERROR
++        mEglSurface = 0;
++
++        eglTerminate(mEglDisplay);
++        EGL_CHECK_ERROR
++        mEglDisplay = 0;
++
++        mActive = false;
++        mVisible = false;
++        mClosed = true;
++    }
++
++    void SwitchEGLWindow::_notifySurfaceCreated(void* window, void* config)
++    {
++        // Set the new window.
++        mWindow = window;
++
++        nwindowSetDimensions((NWindow*)mWindow, mWidth, mHeight);
```

Comment:
This is creating a 1920x1080 window (which is correct), but it is not handling docked/undocked state changes (and adjusting the crop/viewport accordingly). I suppose this will be handled in the future?

Reply:
I've tried to get the window size to switch when changing to and from docked mode and haven't been successful yet. Most attempts involve a crash unless I destroy the GL context, surface, display, etc. The problem one is the GL context and all the things lost when it goes away. Perhaps I'm missing some API that would help. So yeah, I gave up on it for the time being.

Reply:
All you need to do is use nwindowSetCrop and glViewport. I'm not familiar at all with ogre, but presumably there's a way to update the viewport. Please refer to the dynamic_resolution example for more details.